src/upgradeConfig.js

```diff
@@ -1,111 +1,92 @@
-const { presets: oldPresets, plugins: oldPlugins } = require('./packageData');
+const {
+  getNewPackageName,
+  getNewPackageOptionsTransformer
+} = require('./packageData');
 
-// TODO: fix all of this
-function changePresets(config, options = {}) {
-  let presets = config.presets;
+class ConfigItem {
+  constructor(rawConfigItem, type) {
+    const initialName = Array.isArray(rawConfigItem)
+      ? rawConfigItem[0]
+      : rawConfigItem;
 
-  if (!Array.isArray(presets) && typeof presets === 'string') {
-    presets = config.presets = config.presets.split(',').map((preset) => preset.trim());
-  }
+    let name = initialName;
 
-  // check if presets are there
-  if (presets) {
-    // assume it's an array
-    for (let i = 0; i < presets.length; i++) {
-      let preset = presets[i];
-      const presetsToReplace = Object.keys(oldPresets);
-
-      // check if it's a preset with options (an array)
-      if (Array.isArray(preset)) {
-        if (preset[0].indexOf('babel-preset') !== 0 && preset[0].indexOf('@babel/') !== 0) {
-          preset[0] = `babel-preset-${preset[0]}`;
-        }
-        if (presetsToReplace.includes(preset[0])) {
-          if (oldPresets[preset[0]]) {
-            preset[0] = oldPresets[preset[0]];
-          } else {
-            presets.splice(i, 1);
-            i--;
-          }
-        }
-      } else {
-        if (preset.indexOf('babel-preset') !== 0 && preset.indexOf('@babel/') !== 0) {
-          preset = `babel-preset-${preset}`;
-        }
-        if (presetsToReplace.includes(preset)) {
-          if (oldPresets[preset]) {
-            presets[i] = oldPresets[preset];
-          } else {
-            presets.splice(i, 1);
-            i--;
-          }
-        }
-      }
+    if (!name.startsWith(`babel-${type}`) && !name.startsWith('@babel/')) {
+      name = `babel-${type}-${name}`;
     }
 
-    if (options.hasFlow && !presets.includes('@babel/preset-flow')) {
-      presets.push('@babel/preset-flow');
-    }
+    const options =
+      (Array.isArray(rawConfigItem) ? rawConfigItem[1] : {}) || {};
+
+    const newName = getNewPackageName(name);
+    const newOptions = getNewPackageOptionsTransformer(name)(options);
+
+    this.name = newName === undefined ? initialName : newName;
+    this.options = newOptions;
+  }
+
+  render() {
```

Comment:
not tied to this method name -- maybe `toRawConfigItem` would be better (based on the naming i've used elsewhere... that i'm also not totally satisfied with)